packages/syrup/src/encode.js

```diff
@@ -0,0 +1,148 @@
+// @ts-check
+
+import { compareByteArrays } from './compare.js';
+
+const LIST_START = '['.charCodeAt(0);
+const LIST_END = ']'.charCodeAt(0);
+const DICT_START = '{'.charCodeAt(0);
+const DICT_END = '}'.charCodeAt(0);
+const DOUBLE = 'D'.charCodeAt(0);
+const TRUE = new Uint8Array(['t'.charCodeAt(0)]);
+const FALSE = new Uint8Array(['f'.charCodeAt(0)]);
+
+const scratch = new ArrayBuffer(16);
+const scratchBytes = new Uint8Array(scratch);
+const scratchData = new DataView(scratch);
+
+const textEncoder = new TextEncoder();
+
+/**
+ * @param {Record<string, any>} object
+ */
+function encodeDict(object) {
+  let byteLength = 2;
+  let index = 0;
+  const indexes = [];
+  const keys = [];
+  const values = [];
+
+  for (const [key, value] of Object.entries(object)) {
+    // Recursion, it's a thing!
+    // eslint-disable-next-line no-use-before-define
+    const keyBytes = encodeSyrup(key);
+    // eslint-disable-next-line no-use-before-define
+    const valueBytes = encodeSyrup(value);
+    byteLength += keyBytes.byteLength + valueBytes.byteLength;
+    indexes.push(index);
+    keys.push(keyBytes);
+    values.push(valueBytes);
+    index += 1;
+  }
+
+  indexes.sort((i, j) =>
+    compareByteArrays(
+      keys[i],
+      keys[j],
+      0,
+      keys[i].byteLength,
+      0,
+      keys[j].byteLength,
+    ),
+  );
+
+  const bytes = new Uint8Array(byteLength);
+  bytes[0] = DICT_START;
+  bytes[byteLength - 1] = DICT_END;
+  let cursor = 1;
+
+  for (index of indexes) {
+    const key = keys[index];
+    const value = values[index];
+
+    bytes.set(key, cursor);
+    cursor += key.byteLength;
+
+    bytes.set(value, cursor);
+    cursor += value.byteLength;
+  }
+  return bytes;
+}
+
+/**
+ * @param {Array<any>} array
+ */
+function encodeArray(array) {
+  let byteLength = 2;
+  const parts = [];
+  for (const value of array) {
+    // Recursion, it's a thing!
+    // eslint-disable-next-line no-use-before-define
+    const part = encodeSyrup(value);
+    byteLength += part.byteLength;
+    parts.push(part);
+  }
+  const bytes = new Uint8Array(byteLength);
+  bytes[0] = LIST_START;
+  bytes[byteLength - 1] = LIST_END;
+  let cursor = 1;
+  for (const part of parts) {
+    bytes.set(part, cursor);
+    cursor += part.byteLength;
+  }
+  return bytes;
+}
+
+/**
+ * @param {any} value
+ * @returns {Uint8Array}
+ */
+export function encodeSyrup(value) {
+  if (typeof value === 'string') {
+    const suffix = textEncoder.encode(value);
+    const prefix = textEncoder.encode(`${suffix.byteLength}"`);
+    const bytes = new Uint8Array(prefix.byteLength + suffix.byteLength);
+    bytes.set(prefix, 0);
+    bytes.set(suffix, prefix.byteLength);
+    return bytes;
+  }
+
+  if (typeof value === 'number') {
+    scratchData.setFloat64(1, value, false); // big end
```

Comment:
Need to canonicalize `-0` and `NaN` on this side too.

Reply:
Appended a commit that addresses this, and also encodes -0.0 as 0.0 and refuses to decode -0.0.